pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/SourcesImpl.java

```diff
@@ -719,4 +728,14 @@ public SourceConfig getSourceInfo(final String tenant,
                 worker().getWorkerConfig().getValidateConnectorConfig());
         return SourceConfigUtils.convert(sourceConfig, sourceDetails);
     }
+
+    private static boolean hasPackageTypePrefix(String destPkgUrl) {
+        return Arrays.stream(PackageType.values()).anyMatch(type -> destPkgUrl.startsWith(type.toString()));
+    }
+
+    private File downloadPackageFile(String packageName) throws IOException, PulsarAdminException {
```

Comment:
Why do we need to download the package?

Reply:
If we provide a pkgUrl for creating or updating the functions, the function worker has some pre-check of the packages 
https://github.com/apache/pulsar/blob/c17253b3b43c36250ba106dd7a17f35f64f546e5/pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/FunctionsImpl.java#L744

Reply:
You need to validate the existence of the package. But I don't think you should download the package.

Reply:
Ideally, we validate the existence is enough. But that depends on we do the packages format check or function check when uploading a package to the package management service. Then we can remove the checks for the functions which are using the package name to register. But for now, we still need to rely on those checks in the function's code.

I looked into the creation of the function. I saw if a function provided with a packageUrl, it will [download the package and save it as a temp file ](https://github.com/apache/pulsar/blob/1c7ca159d402dcf1a9865410f4872a5741630609/pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/SourcesImpl.java#L156) and do some checks with it. And also it always passes a `File` as the parameter for the method call.
https://github.com/apache/pulsar/blob/1c7ca159d402dcf1a9865410f4872a5741630609/pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/SourcesImpl.java#L160
https://github.com/apache/pulsar/blob/1c7ca159d402dcf1a9865410f4872a5741630609/pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/SourceConfigUtils.java#L244

If we don't download the packages files, we may need to change the existing register function logic.

Reply:
@sijie  Please take a look when you have time. Thanks.

Reply:
@zymap The purpose of introducing a package registry is to avoid checking the packages every time when you create a function using this package. If we need to validate the function package in the package registry, let's add the functionality there.

Reply:
@sijie ok. I open issue #9081 for adding the validation when uploading a package and will do it in another PR.
Can we merge this PR first? Because this PR will allow us to use the package name to create functions.